File: client/lib/device.go
// TODO:
// * Verify success/failure of commands

package dvapbridge

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"log"
	"time"
)
import goserial "github.com/tarm/goserial"

const DVAP_BAUD = 230400

const DVAP_MSG_HOST_SET_CTRL       = 0x00
const DVAP_MSG_HOST_REQ_CTRL_ITEM  = 0x01
const DVAP_MSG_HOST_REQ_CTRL_RANGE = 0x02
const DVAP_MSG_HOST_DATA_ACK       = 0x03
const DVAP_MSG_HOST_DATA_ITEM_0    = 0x04
const DVAP_MSG_HOST_DATA_ITEM_1    = 0x05
const DVAP_MSG_HOST_DATA_ITEM_2    = 0x06
const DVAP_MSG_HOST_DATA_ITEM_3    = 0x07

const DVAP_MSG_TARGET_ITEM_RESPONSE  = 0x00
const DVAP_MSG_TARGET_UNSOLICITED    = 0x01
const DVAP_MSG_TARGET_RANGE_RESPONSE = 0x02
const DVAP_MSG_TARGET_DATA_ACK       = 0x03
const DVAP_MSG_TARGET_DATA_ITEM_0    = 0x04
const DVAP_MSG_TARGET_DATA_ITEM_1    = 0x05
const DVAP_MSG_TARGET_DATA_ITEM_2    = 0x06
const DVAP_MSG_TARGET_DATA_ITEM_3    = 0x07

const DVAP_CTRL_TARGET_NAME        = 0x0001
const DVAP_CTRL_TARGET_SERIAL      = 0x0002
const DVAP_CTRL_IFACE_VER          = 0x0003
const DVAP_CTRL_HW_VER             = 0x0004
const DVAP_CTRL_STATUS             = 0x0005
const DVAP_CTRL_RUN_STATE          = 0x0018
const DVAP_CTRL_MODULATION_TYPE    = 0x0028
const DVAP_CTRL_OPERATION_MODE     = 0x002A
const DVAP_CTRL_SQUELCH_THRESH     = 0x0080
const DVAP_CTRL_OPERATIONAL_STATUS = 0x0090
const DVAP_CTRL_PTT_STATE          = 0x0118
const DVAP_CTRL_EXTERNAL_TR_CTRL   = 0x011A
const DVAP_CTRL_LED_CONTROL        = 0x011C
const DVAP_CTRL_RX_FREQ            = 0x0020
const DVAP_CTRL_TX_FREQ            = 0x0120
const DVAP_CTRL_TX_RX_FREQ         = 0x0220
const DVAP_CTRL_READ_TX_FREQ_LIM   = 0x0230
const DVAP_CTRL_TX_POWER           = 0x0138
const DVAP_CTRL_FREQ_CALIB         = 0x0400
const DVAP_CTRL_BAND_SCAN          = 0x0404
const DVAP_CTRL_DTMF_MSG           = 0x0406

const DVAP_STATUS_STOPPED          = 0x00
const DVAP_STATUS_RUNNING          = 0x01
const DVAP_STATUS_BOOT_IDLE        = 0x0E
const DVAP_STATUS_BOOT_PROG        = 0x0F
const DVAP_STATUS_BOOT_PROG_ERR    = 0x80

const DVAP_RUN_STATE_STOP          = 0x00
const DVAP_RUN_STATE_RUN           = 0x01

const DVAP_MODULATION_FM           = 0x00
const DVAP_MODULATION_GMSK         = 0x01

const DVAP_OPERATION_NORMAL        = 0x00
const DVAP_OPERATION_CWTEST        = 0x01
const DVAP_OPERATION_DEVIATIONTEST = 0x02

const DVAP_PTT_RX_ACTIVE           = 0x00
const DVAP_PTT_TX_ACTIVE           = 0x01

const DVAP_DATA_FM_HDR             = 0x4281
const DVAP_DATA_GMSK_HDR           = 0x2FA0
const DVAP_GMSK_TX_ACK_HDR         = 0x2F60

const DVAP_MSG_MAX_BYTES           = 8191

const DVAP_SQUELCH_MIN             = -128
const DVAP_SQUELCH_MAX             = -45

const DVAP_LED_INTENSITY_MIN       = 0
const DVAP_LED_INTENSITY_MAX       = 100

const DVAP_TX_POWER_MIN            = -12
const DVAP_TX_POWER_MAX            = 10

const DVAP_BAND_SCAN_STEPS_MIN     = 1
const DVAP_BAND_SCAN_STEPS_MAX     = 800
const DVAP_BAND_SCAN_STRIDE_MIN    = 1
const DVAP_BAND_SCAN_STRIDE_MAX    = 255
const DVAP_BAND_SCAN_FREQ_MIN      = 144000000
const DVAP_BAND_SCAN_FREQ_MAX      = 148000000

type RxMessage struct {
	Msgtype byte
	Msg     []byte
}

type Config struct {
	Device string
	serial io.ReadWriteCloser

	txLoopEnabled bool
	txChannel     chan []byte

	rxLoopEnabled bool
	rxChannel     chan RxMessage
	
	// Channel for unsolicited messages
	MsgChannel    chan RxMessage
}

func (c *Config) Open() error {
	var err error
	s := &goserial.Config{Name: c.Device, Baud: DVAP_BAUD}
	c.serial, err = goserial.OpenPort(s)

	if err == nil {
		c.txLoopEnabled = true
		c.txChannel = make(chan []byte)
		go c.serialTxLoop()

		c.rxLoopEnabled = true
		c.rxChannel = make(chan RxMessage)
		go c.serialRxLoop()
	}

	return err
}

func (c *Config) Close() error {
	c.txLoopEnabled = false
	c.rxLoopEnabled = false
	return c.serial.Close()
}

func (c *Config) GetName() string {
	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_TARGET_NAME, []byte{})
	rx := <-c.rxChannel
	return string(rx.Msg[2:])
}

func (c *Config) GetSerial() string {
	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_TARGET_SERIAL, []byte{})
	rx := <-c.rxChannel
	return string(rx.Msg[2:])
}

func (c *Config) GetInterfaceVersion() float32 {
	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_IFACE_VER, []byte{})
	rx := <-c.rxChannel
	return float32(binary.LittleEndian.Uint16(rx.Msg[2:])) / 100.0
}

func (c *Config) GetFirmwareVersions() (bootcodeVer float32, firmwareVer float32) {
	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_HW_VER, []byte{0x00})
	rx := <-c.rxChannel
	bootcodeVer = float32(binary.LittleEndian.Uint16(rx.Msg[3:])) / 100.0

	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_HW_VER, []byte{0x01})
	rx = <-c.rxChannel
	firmwareVer = float32(binary.LittleEndian.Uint16(rx.Msg[3:])) / 100.0
	return
}

func (c *Config) GetStatusCode() byte {
	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_STATUS, []byte{})
	rx := <-c.rxChannel
	return rx.Msg[2]
}

func (c *Config) SetRunState(state byte) {
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_RUN_STATE, []byte{state})
	<-c.rxChannel
}

// NOTE: Can only be sent when DVAP is stopped
func (c *Config) SetModulationType(modulation byte) {
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_MODULATION_TYPE, []byte{modulation})
	<-c.rxChannel
}

func (c *Config) SetOperationMode(mode byte) {
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_OPERATION_MODE, []byte{mode})
	<-c.rxChannel
}

func (c *Config) SetSquelchThreshold(dbm int) {
	squelch := dbm
	if dbm < DVAP_SQUELCH_MIN {
		squelch = DVAP_SQUELCH_MIN
	}
	if dbm > DVAP_SQUELCH_MAX {
		squelch = DVAP_SQUELCH_MAX
	}
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_SQUELCH_THRESH, []byte{byte(squelch & 0xFF)})
	<-c.rxChannel
}

func (c *Config) SetExternalTRControlMode(mode byte) {
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_EXTERNAL_TR_CTRL, []byte{mode})
	<-c.rxChannel
}

func (c *Config) SetLEDControlDVAP() {
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_LED_CONTROL, []byte{0x00, 0x00, 0x00, 0x00})
	<-c.rxChannel
}

// NOTE: Valid intensity values are 0 to 100 (0 = off, 100 = max)
func (c *Config) SetLEDControlHost(yellow uint, red uint, green uint) {
	var y, r, g byte
	if yellow < DVAP_LED_INTENSITY_MIN {
		yellow = DVAP_LED_INTENSITY_MIN
	}
	if yellow > DVAP_LED_INTENSITY_MAX {
		yellow = DVAP_LED_INTENSITY_MAX
	}
	if red < DVAP_LED_INTENSITY_MIN {
		red = DVAP_LED_INTENSITY_MIN
	}
	if red > DVAP_LED_INTENSITY_MAX {
		red = DVAP_LED_INTENSITY_MAX
	}
	if green < DVAP_LED_INTENSITY_MIN {
		green = DVAP_LED_INTENSITY_MIN
	}
	if green > DVAP_LED_INTENSITY_MAX {
		green = DVAP_LED_INTENSITY_MAX
	}
	y = byte(yellow & 0xFF)
	r = byte(red & 0xFF)
	g = byte(green & 0xFF)
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_LED_CONTROL, []byte{0x01, y, r, g})
	<-c.rxChannel
}

func (c *Config) SetRxFrequency(frequency uint32) {
	freq := make([]byte, 4)
	binary.LittleEndian.PutUint32(freq, frequency)
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_RX_FREQ, freq)
	<-c.rxChannel
}

func (c *Config) SetTxFrequency(frequency uint32) {
	freq := make([]byte, 4)
	binary.LittleEndian.PutUint32(freq, frequency)
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_TX_FREQ, freq)
	<-c.rxChannel
}

func (c *Config) SetRxTxFrequency(frequency uint32) {
	freq := make([]byte, 4)
	binary.LittleEndian.PutUint32(freq, frequency)
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_TX_RX_FREQ, freq)
	<-c.rxChannel
}

func (c *Config) GetTxFrequencyLimits() (lower uint32, upper uint32) {
	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_READ_TX_FREQ_LIM, []byte{})
	rx := <-c.rxChannel
	lower = binary.LittleEndian.Uint32(rx.Msg[2:])
	upper = binary.LittleEndian.Uint32(rx.Msg[6:])
	return
}

// Valid range is -12 dBm to 10 dBm
func (c *Config) SetTxPower(dbm int) {
	power := dbm
	if dbm < DVAP_TX_POWER_MIN {
		power = DVAP_TX_POWER_MIN
	}
	if dbm > DVAP_TX_POWER_MAX {
		power = DVAP_TX_POWER_MAX
	}

	pwr := make([]byte, 2)
	binary.LittleEndian.PutUint16(pwr, uint16(power))
	c.sendControlCommand(DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_TX_POWER, pwr)
	<-c.rxChannel
}

// Valid number of steps is 1 to 800
// Valid step size is 100Hz to 25,500Hz in 100 Hz increments (1, 255)
// Valid start frequency range is 144,000,000 to 148,000,000
func (c *Config) GetBandScan(numSteps uint16, stepSize byte, startFreq uint32) (rssi []byte) {
	if numSteps < DVAP_BAND_SCAN_STEPS_MIN {
		numSteps = DVAP_BAND_SCAN_STEPS_MIN
	}
	if numSteps > DVAP_BAND_SCAN_STEPS_MAX {
		numSteps = DVAP_BAND_SCAN_STEPS_MAX
	}
	if stepSize < DVAP_BAND_SCAN_STRIDE_MIN {
		stepSize = DVAP_BAND_SCAN_STRIDE_MIN
	}
	if stepSize > DVAP_BAND_SCAN_STRIDE_MAX {
		stepSize = DVAP_BAND_SCAN_STRIDE_MAX
	}
	if startFreq < DVAP_BAND_SCAN_FREQ_MIN {
		startFreq = DVAP_BAND_SCAN_FREQ_MIN
	}
	if startFreq > DVAP_BAND_SCAN_FREQ_MAX {
		startFreq = DVAP_BAND_SCAN_FREQ_MAX
	}
	
	payload := new(bytes.Buffer)
	binary.Write(payload, binary.LittleEndian, numSteps)
	binary.Write(payload, binary.LittleEndian, stepSize)
	binary.Write(payload, binary.LittleEndian, startFreq)
	c.sendControlCommand(DVAP_MSG_HOST_REQ_CTRL_ITEM, DVAP_CTRL_BAND_SCAN, payload.Bytes())
	rx := <-c.rxChannel
	rssi = rx.Msg[2:]
	return
}

func (c *Config) ParseDTMF() {

}

func (c *Config) serialTx(msg []byte) {
	msgBytes := len(msg)
	var sentBytes int = 0

	// Loop until all data is sent
	for sentBytes < msgBytes {
		n, err := c.serial.Write(msg[sentBytes:msgBytes])
		if n == 0 || err != nil {
			log.Panic(err)
			break
		}
		sentBytes += n
	}
	fmt.Printf("tx: % X\n", msg)
}

func (c *Config) serialTxLoop() {
	for c.txLoopEnabled {
		select {
		case msg := <-c.txChannel:
			c.serialTx(msg)
		case <-time.After(time.Second * 3):
			c.serialTx([]byte{0x03, 0x60, 0x00})
		}
	}
}

func (c *Config) serialRxLoop() {
	buf := make([]byte, DVAP_MSG_MAX_BYTES)

	for c.rxLoopEnabled {
		var receivedBytes int = 0
		var expectedBytes int = 2

		// Receive header
		for receivedBytes < expectedBytes {
			n, err := c.serial.Read(buf[receivedBytes:])
			if n == 0 || err != nil {
				log.Fatal(err)
			}
			receivedBytes += n
		}

		// Receive rest of packet
		expectedBytes = int(buf[0]) + int((buf[1] & 0x1F) << 8)
		for receivedBytes < expectedBytes {
			n, err := c.serial.Read(buf[receivedBytes:expectedBytes])
			if n == 0 || err != nil {
				log.Fatal(err)
			}
			receivedBytes += n
		}

		// Send to channel
		fmt.Printf("rx: % X", buf[:expectedBytes])
		var response RxMessage
		response.Msgtype = byte((buf[1] & 0xE0) >> 5)
		response.Msg = buf[2:expectedBytes]
		fmt.Printf(", type: %d\n", response.Msgtype)

		if response.Msgtype == DVAP_MSG_TARGET_UNSOLICITED {
			c.MsgChannel <- response
		} else {
			c.rxChannel <- response
		}
	}
}

func (c *Config) sendControlCommand(msgtype byte, command uint16, payload []byte) {
	if len(payload) > DVAP_MSG_MAX_BYTES-4 {
		log.Fatal("Payload too long")
	}

	pkt := new(bytes.Buffer)

	// Write dummy header
	header := []byte{0x00, 0x00}
	_, err := pkt.Write(header)
	if err != nil {
		log.Fatal(err)
	}

	// Write command
	err = binary.Write(pkt, binary.LittleEndian, command)
	if err != nil {
		log.Fatal(err)
	}

	// Write payload
	for _, v := range payload {
		err = binary.Write(pkt, binary.LittleEndian, v)
		if err != nil {
			log.Fatal(err)
		}
	}

	// Update header
	var pktlen int = 4 + len(payload)
	pkt.Bytes()[0] = byte(pktlen & 0xFF)
	pkt.Bytes()[1] = (msgtype << 5) | byte((pktlen & 0x1F00) >> 8)

	c.txChannel <- pkt.Bytes()
}

/*
Control message [16 bit header | 16 bit control item | payload]
Data item message [16 bit header | data bytes]
header [8 bit length lsb | 3 bit type | 5 bit length msb]
*/

File: client/device.c
#include <pthread.h>
#include <stdio.h>
#include <sys/select.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "common.h"
#include "device.h"
#include "serial.h"

int
get_name(device_t* ctx, char* name, int name_len)
{
  int sent, ret;
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  sent = dvap_write(ctx, DVAP_MSG_HOST_REQ_CTRL_ITEM,
                    DVAP_CTRL_TARGET_NAME, NULL, 0);
  if (sent <= 0) {
    debug_print("%s\n", "get_name: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);
  ret -= 2;
  ret = (ret <= name_len) ? ret : name_len;
  strncpy(name, (const char *)&buf[2], ret);

  return TRUE;
}

int
set_run_state(device_t* ctx, char state)
{
  int sent, ret;
  unsigned char payload[1];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  payload[0] = state;
  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_RUN_STATE,
                    payload, 1);
  if (sent <= 0) {
    debug_print("%s\n", "set_run_state: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
set_modulation_type(device_t* ctx, char modulation)
{
  int sent, ret;
  unsigned char payload[1];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  payload[0] = modulation;
  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_MODULATION_TYPE,
                    payload, 1);
  if (sent <= 0) {
    debug_print("%s\n", "set_modulation_type: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
set_operation_mode(device_t* ctx, char mode)
{
  int sent, ret;
  unsigned char payload[1];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  payload[0] = mode;
  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_OPERATION_MODE,
                    payload, 1);
  if (sent <= 0) {
    debug_print("%s\n", "set_operation_mode: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
set_squelch_threshold(device_t* ctx, int dbm)
{
  int sent, ret;
  int squelch = dbm;
  unsigned char payload[1];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  squelch = (dbm < DVAP_SQUELCH_MIN) ? DVAP_SQUELCH_MIN : squelch;
  squelch = (dbm < DVAP_SQUELCH_MAX) ? DVAP_SQUELCH_MAX : squelch;

  payload[0] = squelch & 0xFF;
  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_SQUELCH_THRESH,
                    payload, 1);
  if (sent <= 0) {
    debug_print("%s\n", "set_squelch_threshold: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
set_rx_frequency(device_t* ctx, unsigned int hz)
{
  int sent, ret;
  unsigned char payload[4];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  payload[0] = hz & 0xFF;
  payload[1] = (hz >> 8) & 0xFF;
  payload[2] = (hz >> 16) & 0xFF;
  payload[3] = (hz >> 24) & 0xFF;

  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_RX_FREQ,
                    payload, 4);
  if (sent <= 0) {
    debug_print("%s\n", "set_rx_frequency: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
set_tx_frequency(device_t* ctx, unsigned int hz)
{
  int sent, ret;
  unsigned char payload[4];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  payload[0] = hz & 0xFF;
  payload[1] = (hz >> 8) & 0xFF;
  payload[2] = (hz >> 16) & 0xFF;
  payload[3] = (hz >> 24) & 0xFF;

  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_TX_FREQ,
                    payload, 4);
  if (sent <= 0) {
    debug_print("%s\n", "set_tx_frequency: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
set_rxtx_frequency(device_t* ctx, unsigned int hz)
{
  int sent, ret;
  unsigned char payload[4];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  payload[0] = hz & 0xFF;
  payload[1] = (hz >> 8) & 0xFF;
  payload[2] = (hz >> 16) & 0xFF;
  payload[3] = (hz >> 24) & 0xFF;

  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_TX_RX_FREQ,
                    payload, 4);
  if (sent <= 0) {
    debug_print("%s\n", "set_rxtx_frequency: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
set_tx_power(device_t* ctx, int dbm)
{
  int sent, ret;
  int power;
  unsigned char payload[2];
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  if (!ctx) return FALSE;

  power = (dbm < DVAP_TX_POWER_MIN) ? DVAP_TX_POWER_MIN : dbm;
  power = (dbm > DVAP_TX_POWER_MAX) ? DVAP_TX_POWER_MAX : dbm;

  payload[0] = power & 0xFF;
  payload[1] = (power >> 8) & 0xFF;

  sent = dvap_write(ctx, DVAP_MSG_HOST_SET_CTRL, DVAP_CTRL_TX_RX_FREQ,
                    payload, 2);
  if (sent <= 0) {
    debug_print("%s\n", "set_tx_power: error on dvap_write");
    return FALSE;
  }

  queue_remove(&(ctx->rxq), buf, &ret);

  return TRUE;
}

int
dvap_init(device_t* ctx, char* portname, dvap_rx_fptr callback)
{
  int fd;
  if (!ctx) return FALSE;

  ctx->callback = callback;

  // serial port
  fd = serial_open(portname, DVAP_BAUD);
  if (fd < 0) {
    return FALSE;
  }
  ctx->fd = fd;

  // shutdown variables
  ctx->shutdown = FALSE;
  pthread_mutex_init(&(ctx->shutdown_mutex), NULL);

  // transmitter variables
  pthread_mutex_init(&(ctx->tx_mutex), NULL);
  pthread_mutex_init(&(ctx->ptt_mutex), NULL);
  ctx->ptt_active = FALSE;

  // receive queue
  queue_init(&(ctx->rxq));
  pthread_create(&(ctx->rx_thread), NULL, dvap_read_loop, ctx);

  return TRUE;
}

int
dvap_start(device_t* ctx)
{
  if (!ctx) return FALSE;

  if (!set_run_state(ctx, DVAP_RUN_STATE_RUN)) {
    fprintf(stderr, "Error setting DVAP run state to RUN\n");
    return FALSE;
  }

  pthread_create(&(ctx->watchdog_thread), NULL, dvap_watchdog_loop, ctx);
  return TRUE;
}

void
dvap_wait(device_t* ctx)
{
  if (!ctx) return;
  pthread_join(ctx->rx_thread, NULL);
  pthread_join(ctx->watchdog_thread, NULL);
  close(ctx->fd);

  pthread_mutex_destroy(&(ctx->shutdown_mutex));
  pthread_mutex_destroy(&(ctx->tx_mutex));
}

int
dvap_stop(device_t* ctx)
{
  if (!ctx) return FALSE;

  if (!set_run_state(ctx, DVAP_RUN_STATE_STOP)) {
    fprintf(stderr, "Error setting DVAP run state to STOP\n");
    return FALSE;
  }
  pthread_mutex_lock(&(ctx->shutdown_mutex));
  ctx->shutdown = TRUE;
  pthread_mutex_unlock(&(ctx->shutdown_mutex));

  return TRUE;
}

int
dvap_pkt_write(device_t* ctx, unsigned char* buf, int buf_bytes)
{
  int n;
  int sent_bytes = 0;

  pthread_mutex_lock(&(ctx->tx_mutex));
  while (sent_bytes < buf_bytes) {
    n = write(ctx->fd, &buf[sent_bytes], buf_bytes-sent_bytes);
    if (DEBUG) {
      printf("[%ld] dvap_pkt_write: %d bytes\n", time(NULL), buf_bytes);
    }
    if (n <= 0) {
      pthread_mutex_unlock(&(ctx->tx_mutex));
      fprintf(stderr, "dvap_pkt_write - error writing to device\n");
      return -1;
    }
    sent_bytes += n;
  }
  pthread_mutex_unlock(&(ctx->tx_mutex));
  return sent_bytes;
}

int
dvap_write(device_t* ctx, char msg_type, int command, unsigned char* payload,
           int payload_bytes)
{
  int n;
  int total_sent_bytes = 0;
  int payload_sent_bytes = 0;
  unsigned char buf[4];
  int pktlen;

  if (!ctx) return -1;

  if (payload_bytes > DVAP_MSG_MAX_BYTES-4) {
    fprintf(stderr, "Error in send_command, payload too long\n");
    return -1;
  }

  pktlen = 4 + payload_bytes;
  buf[0] = pktlen & 0xFF;
  buf[1] = (msg_type << 5) | ((pktlen & 0x1F00) >> 8);

  buf[2] = command & 0xFF;
  buf[3] = (command >> 8) & 0xFF;

  if (DEBUG) {
    hex_dump("tx", buf, 4);
  }

  pthread_mutex_lock(&(ctx->tx_mutex));
  while(total_sent_bytes < 4) {
    n = write(ctx->fd, &buf[total_sent_bytes], 4-total_sent_bytes);
    if (n <= 0) {
      pthread_mutex_unlock(&(ctx->tx_mutex));
      debug_print("dvap_write - returned %d\n", n);
      return n;
    }
    total_sent_bytes += n;
  }

  if (payload) {
    if (DEBUG) {
      hex_dump("tx payload", payload, payload_bytes);
    }
    while(payload_sent_bytes < payload_bytes) {
      n = write(ctx->fd, &payload[payload_sent_bytes],
                payload_bytes-payload_sent_bytes);
      if (n <= 0) {
        pthread_mutex_unlock(&(ctx->tx_mutex));
        debug_print("dvap_write - returned %d\n", n);
        return n;
      }
      payload_sent_bytes += n;
      total_sent_bytes += n;
    }
  }
  pthread_mutex_unlock(&(ctx->tx_mutex));
  return total_sent_bytes;
}

int
dvap_read(device_t* ctx, char* msg_type, unsigned char* buf, int buf_bytes)
{
  if (!ctx) return -1;
  return packet_read(ctx->fd, msg_type, buf, buf_bytes);
}

int
dvap_should_shutdown(device_t* ctx)
{
  int shutdown;
  if (!ctx) return TRUE;
  
  pthread_mutex_lock(&(ctx->shutdown_mutex));
  shutdown = ctx->shutdown;
  pthread_mutex_unlock(&(ctx->shutdown_mutex));
  return shutdown;
}

void*
dvap_watchdog_loop(void* arg)
{
  device_t* ctx = (device_t *)arg;
  unsigned char buf[3];
  int ptt_active;
  int counter = DVAP_WATCHDOG_SECS;
  buf[0] = 0x03;
  buf[1] = 0x60;
  buf[2] = 0x00;

  while (!dvap_should_shutdown(ctx)) {
    if (counter <= 0) {
      pthread_mutex_lock(&(ctx->ptt_mutex));
      ptt_active = ctx->ptt_active;
      pthread_mutex_unlock(&(ctx->ptt_mutex));

      // Only send watchdog keepalive message if transmitter is not in use
      if (!ptt_active) {
        pthread_mutex_lock(&(ctx->tx_mutex));
        write(ctx->fd, buf, 3);
        pthread_mutex_unlock(&(ctx->tx_mutex));
        counter = DVAP_WATCHDOG_SECS;
        if (DEBUG) {
          hex_dump("watchdog tx", buf, 3);
        }
      }
    }
    counter -= 1;
    sleep(1);
  }

  return NULL;
}

void* 
dvap_read_loop(void* arg)
{
  device_t* ctx = (device_t *)arg;

  fd_set set;
  char msg_type;
  int ret;
  unsigned char buf[DVAP_MSG_MAX_BYTES];

  struct timeval timeout;
  while(!dvap_should_shutdown(ctx)) {
    // Check if data is available for reading
    // Timeout and try again if nothing is available
    FD_ZERO(&set);
    FD_SET(ctx->fd, &set);
    // Linux version of select overwrites timeout, so we set it on each
    // iteration of the while loop
    timeout.tv_sec = 0;
    timeout.tv_usec = DVAP_READ_TIMEOUT_USEC;
    ret = select(ctx->fd+1, &set, NULL, NULL, &timeout);
    if (ret < 0) {
      fprintf(stderr, "Error waiting for data from DVAP\n");
      return NULL;
    }
    else if (ret == 0) {
      //debug_print("%s\n", "DVAP select timeout");
      continue;
    }

    // Read packet
    ret = dvap_read(ctx, &msg_type, buf, DVAP_MSG_MAX_BYTES);
    if (ret < 0) {
      fprintf(stderr, "Error reading from DVAP\n");
      return NULL;
    }
    else if (ret == 0) {
      fprintf(stderr, "Timeout while reading from DVAP\n");
      return NULL;
    }

    // Call appropriate handler depending on message type
    switch (msg_type) {

    // Response to a host initiated request
    case DVAP_MSG_TARGET_ITEM_RESPONSE:
    case DVAP_MSG_TARGET_RANGE_RESPONSE:
      queue_insert(&(ctx->rxq), &buf[2], ret-2);
      if (DEBUG) {
        hex_dump("rx", buf, ret);
      }
      break;

    // Status message
    case DVAP_MSG_TARGET_UNSOLICITED:
      dvap_parse_rx_unsolicited(ctx, &buf[2], ret-2);
      break;

    // Ignore target data acks for now
    case DVAP_MSG_TARGET_DATA_ACK:
      break;

    // Radio data
    case DVAP_MSG_TARGET_DATA_ITEM_0:
    case DVAP_MSG_TARGET_DATA_ITEM_1:
    case DVAP_MSG_TARGET_DATA_ITEM_2:
    case DVAP_MSG_TARGET_DATA_ITEM_3:
      (ctx->callback)(buf, ret);
      break;

    default:
      fprintf(stderr, "rx: unrecognized response type: %d\n", msg_type);
      break;
    }
  }

  return NULL;
}

void
dvap_parse_rx_unsolicited(device_t* ctx, unsigned char* buf, int buf_len)
{
  int ctrl_code = (buf[1] << 8) + buf[0];

  switch (ctrl_code) {
  case DVAP_CTRL_OPERATIONAL_STATUS:
    //dvap_print_operational_status(buf, buf_len);
    break;
  case DVAP_CTRL_PTT_STATE:
    //dvap_print_ptt_state(buf, buf_len);
    pthread_mutex_lock(&(ctx->ptt_mutex));
    if (buf_len >= 3) {
      if (buf[2] <= 0) {
        ctx->ptt_active = FALSE;
      }
      else {
        ctx->ptt_active = TRUE;
      }
    }
    pthread_mutex_unlock(&(ctx->ptt_mutex));
    break;
  case DVAP_CTRL_DTMF_MSG:
    dvap_print_dtmf(buf, buf_len);
    break;
  default:
    if (DEBUG) {
      hex_dump("rx unsolicited other", buf, buf_len);
    }
    break;
  }
}

void
dvap_print_operational_status(unsigned char* buf, int buf_len)
{
  char rssi, squelch, fifo_free;
  if (buf_len < 5) return;

  rssi = buf[2];
  squelch = buf[3];
  fifo_free = buf[4];

  printf("rssi: %04d, squelch: %d, tx fifo free: %03d\n", rssi, squelch,
         fifo_free);
}

void
dvap_print_ptt_state(unsigned char* buf, int buf_len)
{
  if (buf_len < 3) return;
  if (buf[2] <= 0) {
    printf("ptt: receive active\n");
  }
  else {
    printf("ptt: transmit active\n");
  }
}

void
dvap_print_dtmf(unsigned char* buf, int buf_len)
{
  if (buf_len < 3) return;
  if (buf[2] == 0) {
    printf("dtmf: release\n");
  }
  else {
    printf("dtmf: %c\n", buf[2]);
  }
}
